extensions/typescript/src/features/jsDocCompletionProvider.ts

```diff
@@ -0,0 +1,184 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+import { Position, Selection, Range, CompletionItemProvider, CompletionItemKind, TextDocument, CancellationToken, CompletionItem, window, commands, Uri, ProviderResult, TextEditor } from 'vscode';
+
+import { ITypescriptServiceClient } from '../typescriptService';
+import { FileLocationRequestArgs, DocCommandTemplateResponse } from '../protocol';
+
+import * as nls from 'vscode-nls';
+const localize = nls.loadMessageBundle();
+
+const tryCompleteJsDocCommand = '_typeScript.tryCompleteJsDoc';
+
+
+class JsDocCompletionItem extends CompletionItem {
+	constructor(file: Uri, position: Position) {
+		super('/** @param */', CompletionItemKind.Snippet);
+		this.detail = localize('typescript.jsDocCompletionItem.detail', 'Complete JSDoc comment');
+		this.insertText = '';
+		this.command = {
+			title: 'Try Complete Js Doc',
+			command: tryCompleteJsDocCommand,
+			arguments: [file, position]
+		};
+	}
+}
+
+export default class JsDocCompletionHelper implements CompletionItemProvider {
+
+	constructor(
+		private client: ITypescriptServiceClient,
+	) {
+		window.onDidChangeTextEditorSelection(e => {
+			if (e.textEditor.document.languageId !== 'typescript'
+				&& e.textEditor.document.languageId !== 'typescriptreact'
+				&& e.textEditor.document.languageId !== 'javascript'
+				&& e.textEditor.document.languageId !== 'javascriptreact'
+			) {
+				return;
+			}
+
+			const selection = e.selections[0];
+			if (!selection.start.isEqual(selection.end)) {
+				return;
+			}
+			if (this.shouldAutoShowJsDocSuggestion(e.textEditor.document, selection.start)) {
+				return commands.executeCommand('editor.action.triggerSuggest');
```

Comment:
We have all sorts of commands and settings to trigger suggestions, for instance 24x7 or provider-specific trigger characters. This is somewhat backwards and I wonder why that is needed? This provider could for instance use `*` as trigger character to complete `/**` sequences. It's also being continuously asked for suggestion because the default config is `editor.quickSuggestion: true`. The idea is not trigger yourself but to know when to return something